### PR TITLE
[ML] Rerank response format change

### PR DESCRIPTION
**Pull request by maxhniebergall**
This PR changes the response format of the Cohere Rerank inference endpoint. Previously, the field values `relevance_score` and `index` were Strings, and now they will be `float` and `int`, respectively. 

Elasticsearch-spec PR to follow.

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferenceNamedWriteablesProvider.java [hunk 1]
```diff
@@ -184,6 +189,9 @@ public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
         namedWriteables.add(
             new NamedWriteableRegistry.Entry(TaskSettings.class, CohereEmbeddingsTaskSettings.NAME, CohereEmbeddingsTaskSettings::new)
         );
+        namedWriteables.add(
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferenceNamedWriteablesProvider.java:192]: I think we need the service settings too right?

```
        namedWriteables.add(
            new NamedWriteableRegistry.Entry(ServiceSettings.class, CohereRerankServiceSettings.NAME, CohereRerankServiceSettings::new)
        );
```

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/cohere/CohereRankedResponseEntity.java [hunk 2]
```diff
@@ -28,6 +28,10 @@
 
 public class CohereRankedResponseEntity {
 
+    private static final org.apache.logging.log4j.Logger logger = org.apache.logging.log4j.LogManager.getLogger(
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/cohere/CohereRankedResponseEntity.java:31]: hmm I wonder if we can have this as an import like in other spots 🤔 so we don't have the class path

```
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
```

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/cohere/CohereRankedResponseEntity.java [hunk 3]
```diff
@@ -142,7 +146,18 @@ private static RankedDocsResults.RankedDoc parseRankedDocObject(XContentParser p
...
+            logger.error("Failed to find required field [index] in Cohere embeddings response");
+        }
+        if (relevanceScore == -1) {
+            logger.error("Failed to find required field [relevance_score] in Cohere embeddings response");
+        }
+        if (documentText == null) {
+            logger.error("Failed to find required field [document] in Cohere embeddings response");
+        }
+
+        return new RankedDocsResults.RankedDoc(index, relevanceScore, String.valueOf(documentText));
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/cohere/CohereRankedResponseEntity.java:160]: Do we need the `String.valueOf(documentText)`? It's already a string right?

Comment (maxhniebergall) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/cohere/CohereRankedResponseEntity.java:160]: I think it was a mistake to use valueOf there.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferenceNamedWriteablesProvider.java (hunk 1 above)
Comment (maxhniebergall) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferenceNamedWriteablesProvider.java:192]: Thanks Jonathan! Do you think we can come up with a list of all the things that need to go in this file and add it as a comment here? I always have trouble figuring out what exactly needs to go in here.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/RankedDocsResults.java [hunk 4]
```diff
@@ -57,11 +102,25 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-            out.writeString(index);
-            out.writeString(relevanceScore);
-            out.writeString(text);
+            if (out.getTransportVersion().onOrAfter(ML_INFERENCE_RERANK_NEW_RESPONSE_FORMAT)) {
+                out.writeInt(index);
+                out.writeFloat(relevanceScore);
+                out.writeString(text);
+            } else {
+                out.writeString(String.valueOf(index));
+                out.writeString(String.valueOf(relevanceScore));
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/RankedDocsResults.java:121]:
```suggestion
                out.writeString(Integer.toString(index));
                out.writeString(Float.toString(relevanceScore));
```

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/cohere/CohereRankedResponseEntity.java (hunk 3 above)
Comment (davidkyle) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/cohere/CohereRankedResponseEntity.java:160]: > I think it was a mistake to use valueOf there.

👍 I left a similar suggestion

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferenceNamedWriteablesProvider.java (hunk 1 above)
Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferenceNamedWriteablesProvider.java:193]: Yeah I believe it's any time we create a class that implements `NamedWriteable`. That kind of hard to know without following the `implements` of classes, but essentially anything that implements `TaskSettings`, `ServiceSettings`, or `InferenceServiceResults` at the moment.

Technically also `InferenceResults` but we shouldn't really be implementing that interface anymore within the inference plugin.

**Review by jonathan-buttner — APPROVED** [review]

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferenceNamedWriteablesProvider.java (hunk 1 above)
Comment (maxhniebergall) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferenceNamedWriteablesProvider.java:193]: Nice, I will add a comment specifying this at the top of this class.

**Pull request merged**